### PR TITLE
Switched rest_utils to use cached Requests.Session objects

### PR DESCRIPTION
**Pull request by jinzhang21**
Signed-off-by: jinzhang21 <jin.zhang@databricks.com>

## What changes are proposed in this pull request?

Currently for every single HTTP request, we create a new Session object, which has to re-establish a new TCP connection to remote server before sending and receiving data. This is inefficient and sometimes can cause client be rate-limited by the server.

## How is this patch tested?

Existing unit tests.
Manual test on Databricks simulated environment:
- Log 1000 metrics
- Log 1000 params
- Log 100 artifacts
- Log 100 models 

Some anecdotal performance improvements are observed:
- From 1.23.1, total run time about **410 seconds**
- From current PR, total run time about **340 seconds**.
Above measurements are repeated twice.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/rest_utils.py
Comment (harupy) [inline mlflow/utils/rest_utils.py:38]: Can we use [functools.lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache)?

File: mlflow/utils/rest_utils.py
Comment (harupy) [inline mlflow/utils/rest_utils.py:96]:
```suggestion
    with _get_request_session(retry_kwargs) as session:
        return session.request(method, url, **kwargs)
```

Can we use `with`?

Comment (harupy) [reply mlflow/utils/rest_utils.py:96]: The reason we don't use `with` is to avoid closing the session?

File: tests/utils/test_rest_utils.py
Comment (harupy) [inline tests/utils/test_rest_utils.py:7]:
```suggestion
from mlflow.utils import rest_utils
```

nit: Can we use from-import?

File: mlflow/utils/rest_utils.py
Comment (harupy) [reply mlflow/utils/rest_utils.py:38]:
```python
>>> hash({"a": 0})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
```

`dict` is

Comment (harupy) [reply mlflow/utils/rest_utils.py:38]: So we need a trick like this if we use `lru_cache`:

```python
from functools import lru_cache


class HashableDict(dict):
    def __hash__(self):
        return hash(frozenset(self.items()))


@lru_cache(maxsize=100)
def func(kwargs):
    print(kwargs)


func(HashableDict({"a": 0}))
func(HashableDict({"a": 0}))
func(HashableDict({"a": 0, "b": 1}))
```

Comment (BenWilson2) [reply mlflow/utils/rest_utils.py]: Won't we have to set a cache expiry time in order to fetch updated page results server-side and prevent stale cached data from being displayed if a previous request's data updates? (I imagine this might be applicable to certain pages and not others, though)

`@lru_cache(maxsize=<x>, seconds=30)` (or some other reasonable time period based on updating periodicities of run data while navigating). 

Would we want to cache Put requests or just Get requests?

Potential future food for thought for this: [requests-cache](https://requests-cache.readthedocs.io/en/latest/user_guide.html) which can give endpoint isolation for caching certain endpoints with different settings and a bunch of other potentially useful features to lessen the tracking server burden.

Comment (jinzhang21) [reply mlflow/utils/rest_utils.py]: Not sure if I fully understand your concern, but the added function LRU only caches the `requests.Session` object. All outgoing requests (initiated by `session.request`) are not cached and should never be.

The objective here is to persist the TCP channel across HTTP requests to the same destination IP/Port combo, nothing else.

File: mlflow/utils/rest_utils.py
Comment (dbczumar) [inline mlflow/utils/rest_utils.py:38]: We use the same max retries, backoff factor, and retry codes for nearly all requests, implying that only a single session will be used. Shall we scope this based on URL as well?

File: mlflow/utils/rest_utils.py
Comment (BenWilson2) [reply mlflow/utils/rest_utils.py]: I was just curious if a NAT router / Firewall change between the server and the client introducing a network change would lead a long-running session into defaulting to maximum backoff retry event waiting before invalidating the cache value and establishing a new TCP connection. Maybe I'm thinking about this wrong though.

File: mlflow/utils/rest_utils.py
Comment (jinzhang21) [reply mlflow/utils/rest_utils.py:38]: It's very much true. And I think it is working as intended -- most of HTTP requests are sent from a single client (current runtime) to a single destination server, which should and could be considered as in the same session.

Session uses urlib3 to manage the connection pool (on TCP level) under the hood. If url differs, a new connection will be created there. I don't see a specific reason to create new sessions for that purpose.

File: mlflow/utils/rest_utils.py
Comment (jinzhang21) [reply mlflow/utils/rest_utils.py]: See the top section of this page: https://docs.python-requests.org/en/latest/user/advanced/#session-objects
I think we can safely leave the network management and TCP connection re-use to urllib3, which is managed by the session object.

File: mlflow/utils/rest_utils.py
Comment (dbczumar) [reply mlflow/utils/rest_utils.py:38]: We make cloud storage requests to S3, Azure, GCP, etc. Is there anything problematic with reusing the same session object across all of these platforms? I'm just wondering why this wouldn't be the default behavior for requests if single session use across an entire application was recommended it most cases.

Comment (jinzhang21) [reply mlflow/utils/rest_utils.py:38]: We can certainly cache sessions based on URLs. My point is that it's less efficient, in the sense that Session is just a thin wrapper used to persist a few settings across requests. More specifically, it "Provides cookie persistence, connection-pooling, and configuration.", according to the class documentation: https://github.com/psf/requests/blob/main/requests/sessions.py#L327

From https://github.com/psf/requests/blob/main/requests/sessions.py#L343, among all stuff a Session persists currently
```
        'headers', 'cookies', 'auth', 'proxies', 'hooks', 'params', 'verify',
        'cert', 'adapters', 'stream', 'trust_env', 'max_redirects',
```
the only setting that's being persisted in our use case is the `adapters` (which could change depending on the retry logic), which does the TCP connection pooling and re-use.

In the future, if we decide to persist auth and cookies, it probably makes sense to change the cache logic to take these into account, which should happen automatically since auth and cookie settings have to be passed into the session generator function `_get_request_session` as part of the argument, which will change the session cache key signature.

File: mlflow/utils/rest_utils.py
Comment (BenWilson2) [reply mlflow/utils/rest_utils.py]: 100%. I forgot about how requests will drop the connection and re-validate with a new connection from the pool if a connection cannot be established.

File: mlflow/utils/rest_utils.py
Comment (dbczumar) [reply mlflow/utils/rest_utils.py:38]: Understood! Thanks for the clarification. Can we manually confirm that logging both models and metadata (metrics, params, etc) works properly on Databricks before merging? I see we've already tested metrics.

**Review by dbczumar — APPROVED** [review]
LGTM pending results of manual testing with model logging + metadata logging. Thanks @jinzhang21 !

File: mlflow/utils/rest_utils.py
Comment (jinzhang21) [reply mlflow/utils/rest_utils.py:38]: Tested on Databricks with logging metrics, params, artifacts and models repeatedly. No problem was observed. We see some performance improvements, especially on repeated small requests such as metric logging.

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**
